File: examples/basic-demos/lib/logger.js
/**
 * Structured logging infrastructure for Demo Node.js CLI.
 * 
 * This module provides structured logging with context management using AsyncLocalStorage.
 * Environment variables:
 * - LOG_LEVEL: Set logging level (debug, info, warn, error) - default: info
 * - LOG_OUTPUT: Set output destination (stdout, stderr, file:<path>) - default: stdout
 * - ENVIRONMENT: Set environment (production/development) - affects format
 */

const { AsyncLocalStorage } = require('async_hooks');
const winston = require('winston');
const fs = require('fs');
const path = require('path');

// Context storage for async context management
const contextStorage = new AsyncLocalStorage();

// Custom formatter that outputs structured logs based on environment
const structuredFormatter = winston.format.printf(({ timestamp, level, message, ...meta }) => {
    const context = contextStorage.getStore() || {};
    const environment = process.env.ENVIRONMENT || 'development';
    const isProduction = environment.toLowerCase() === 'production' || environment.toLowerCase() === 'prod';
    
    const logData = {
        timestamp,
        level,
        message,
        ...meta
    };
    
    // Add context if available
    if (Object.keys(context).length > 0) {
        logData.context = context;
    }
    
    if (isProduction) {
        // JSON format for production
        return JSON.stringify(logData);
    } else {
        // Human-readable format for development
        const contextStr = Object.keys(context).length > 0 
            ? ` [${Object.entries(context).map(([k, v]) => `${k}=${v}`).join(', ')}]` 
            : '';
        
        const extraFields = Object.keys(meta).filter(key => 
            !['timestamp', 'level', 'message', 'context'].includes(key)
        );
        const extraStr = extraFields.length > 0 
            ? ` ${JSON.stringify(extraFields.reduce((acc, key) => ({ ...acc, [key]: meta[key] }), {}))}` 
            : '';
        
        return `${timestamp} ${level.toUpperCase().padEnd(8)} Demo Node.js CLI.${message}${contextStr}${extraStr}`;
    }
});

// Create logger instance
let logger = null;

/**
 * Initialize structured logging for Demo Node.js CLI.
 * 
 * Environment Variables:
 * - LOG_LEVEL: Set logging level (debug, info, warn, error) - default: info
 * - LOG_OUTPUT: Set output destination (stdout, stderr, file:<path>) - default: stdout
 * - ENVIRONMENT: Set environment (production/development) - affects format
 */
function setupLogging() {
    const logLevel = process.env.LOG_LEVEL || 'info';
    const logOutput = process.env.LOG_OUTPUT || 'stdout';
    
    const transports = [];
    
    if (logOutput === 'stderr') {
        // All logs to stderr
        transports.push(new winston.transports.Console({ 
            stderrLevels: ['error', 'warn', 'info', 'debug'],
            format: winston.format.combine(
                winston.format.timestamp(),
                structuredFormatter
            )
        }));
    } else if (logOutput.startsWith('file:')) {
        // Log to file
        const logFile = logOutput.substring(5); // Remove 'file:' prefix
        const logDir = path.dirname(logFile);
        
        // Ensure log directory exists
        if (!fs.existsSync(logDir)) {
            fs.mkdirSync(logDir, { recursive: true });
        }
        
        transports.push(new winston.transports.File({
            filename: logFile,
            format: winston.format.combine(
                winston.format.timestamp(),
                structuredFormatter
            )
        }));
    } else if (logOutput === 'stdout') {
        // Container-friendly: info/debug to stdout, warn/error to stderr
        transports.push(new winston.transports.Console({
            level: 'info',
            stderrLevels: ['error', 'warn'],
            format: winston.format.combine(
                winston.format.timestamp(),
                structuredFormatter
            )
        }));
    } else {
        // Default to stdout for unknown options
        transports.push(new winston.transports.Console({
            format: winston.format.combine(
                winston.format.timestamp(),
                structuredFormatter
            )
        }));
    }
    
    logger = winston.createLogger({
        level: logLevel,
        transports,
        exitOnError: false
    });
    
    // Log startup message
    logger.info(`Logging initialized: level=${logLevel}, output=${logOutput}`);
}

/**
 * Get a logger instance.
 * 
 * @param {string} name - Logger name (module name)
 * @returns {Object} Winston logger instance with context support
 */
function getLogger(name = 'main') {
    if (!logger) {
        setupLogging();
    }
    
    // Return logger with context-aware methods
    return {
        debug: (message, meta = {}) => logger.debug(message, { module: name, ...meta }),
        info: (message, meta = {}) => logger.info(message, { module: name, ...meta }),
        warn: (message, meta = {}) => logger.warn(message, { module: name, ...meta }),
        error: (message, meta = {}) => logger.error(message, { module: name, ...meta }),
        
        // Context management methods
        setContext: setContext,
        clearContext: clearContext,
        updateContext: updateContext,
        getContext: getContext,
        removeContextKeys: removeContextKeys
    };
}

/**
 * Set logging context variables that will be included in all log messages within the current async context.
 * 
 * @param {Object} contextData - Key-value pairs to add to logging context
 * @param {Function} callback - Function to execute with the context
 * 
 * @example
 * setContext({ operationId: 'op_123', user: 'admin' }, () => {
 *     logger.info('Operation started'); // Will include context automatically
 * });
 */
function setContext(contextData, callback) {
    if (typeof contextData === 'function' && !callback) {
        // If only one function is provided, treat it as callback with empty context
        callback = contextData;
        contextData = {};
    }
    
    const currentContext = contextStorage.getStore() || {};
    const newContext = { ...currentContext, ...contextData };
    
    if (callback) {
        return contextStorage.run(newContext, callback);
    } else {
        // For synchronous context setting (less recommended)
        return contextStorage.run(newContext, () => {});
    }
}

/**
 * Clear all logging context variables.
 */
function clearContext() {
    return contextStorage.run({}, () => {});
}

/**
 * Update existing context with new values.
 * 
 * @param {Object} contextData - Key-value pairs to update in logging context
 */
function updateContext(contextData) {
    const currentContext = contextStorage.getStore() || {};
    const updatedContext = { ...currentContext, ...contextData };
    return contextStorage.run(updatedContext, () => {});
}

/**
 * Get current logging context.
 * 
 * @returns {Object} Current context object
 */
function getContext() {
    return { ...(contextStorage.getStore() || {}) };
}

/**
 * Remove specific keys from logging context.
 * 
 * @param {...string} keys - Context keys to remove
 */
function removeContextKeys(...keys) {
    const currentContext = contextStorage.getStore() || {};
    const updatedContext = { ...currentContext };
    
    keys.forEach(key => {
        delete updatedContext[key];
    });
    
    return contextStorage.run(updatedContext, () => {});
}

// Export the logging interface
module.exports = {
    setupLogging,
    getLogger,
    setContext,
    clearContext,
    updateContext,
    getContext,
    removeContextKeys
};

File: src/examples/validation/test_universal_python_output/src/test_universal_cli/logger.py
"""
Structured logging infrastructure for Test Universal CLI.

This module provides structured logging with context management for enhanced observability.
Environment variables:
- LOG_LEVEL: Set logging level (DEBUG, INFO, WARNING, ERROR) - default: INFO
- LOG_OUTPUT: Set output destination (stdout, stderr, file:<path>) - default: stdout
- ENVIRONMENT: Set environment (production/development) - affects format
"""

import json
import logging
import os
import sys
from contextvars import ContextVar
from typing import Dict, Any, Optional
from pathlib import Path

# Context variables for structured logging
_log_context: ContextVar[Dict[str, Any]] = ContextVar("log_context", default={})


class StructuredFormatter(logging.Formatter):
    """Custom formatter that outputs structured logs based on environment."""

    def __init__(self):
        super().__init__()
        self.environment = os.getenv("ENVIRONMENT", "development")
        self.is_production = self.environment.lower() in ("production", "prod")

    def format(self, record: logging.LogRecord) -> str:
        """Format log record as JSON for production or readable format for development."""

        # Get current context
        context = _log_context.get({})

        # Build log data
        log_data = {
            "timestamp": self.formatTime(record),
            "level": record.levelname,
            "logger": record.name,
            "message": record.getMessage(),
            "module": record.module,
            "function": record.funcName,
            "line": record.lineno,
        }

        # Add context if available
        if context:
            log_data["context"] = context

        # Add exception info if present
        if record.exc_info:
            log_data["exception"] = {
                "type": record.exc_info[0].__name__ if record.exc_info[0] else None,
                "message": str(record.exc_info[1]) if record.exc_info[1] else None,
                "traceback": (
                    self.formatException(record.exc_info) if record.exc_info else None
                ),
            }

        # Add extra fields from log record
        extra_fields = {}
        for key, value in record.__dict__.items():
            if key not in [
                "name",
                "msg",
                "args",
                "levelname",
                "levelno",
                "pathname",
                "filename",
                "module",
                "lineno",
                "funcName",
                "created",
                "msecs",
                "relativeCreated",
                "thread",
                "threadName",
                "processName",
                "process",
                "message",
                "exc_info",
                "exc_text",
                "stack_info",
            ]:
                extra_fields[key] = value

        if extra_fields:
            log_data["extra"] = extra_fields

        if self.is_production:
            # JSON format for production
            return json.dumps(log_data)
        else:
            # Human-readable format for development
            context_str = (
                f" [{', '.join(f'{k}={v}' for k, v in context.items())}]"
                if context
                else ""
            )
            extra_str = f" {extra_fields}" if extra_fields else ""

            base_msg = f"{log_data['timestamp']} {log_data['level']:8} {log_data['logger']:20} {log_data['message']}"

            if context_str or extra_str:
                base_msg += f"{context_str}{extra_str}"

            # Add exception info if present
            if record.exc_info:
                base_msg += f"\n{self.formatException(record.exc_info)}"

            return base_msg


def setup_logging() -> None:
    """
    Initialize structured logging for Test Universal CLI.

    Environment Variables:
    - LOG_LEVEL: Set logging level (DEBUG, INFO, WARNING, ERROR) - default: INFO
    - LOG_OUTPUT: Set output destination (stdout, stderr, file:<path>) - default: stdout
    - ENVIRONMENT: Set environment (production/development) - affects format
    """

    # Get configuration from environment
    log_level = os.getenv("LOG_LEVEL", "INFO").upper()
    log_output = os.getenv("LOG_OUTPUT", "stdout").lower()

    # Configure logging level
    try:
        level = getattr(logging, log_level)
    except AttributeError:
        level = logging.INFO
        print(f"Warning: Invalid LOG_LEVEL '{log_level}', using INFO", file=sys.stderr)

    # Set up formatter
    formatter = StructuredFormatter()

    # Configure handlers based on LOG_OUTPUT
    handlers = []

    if log_output == "stderr":
        # All logs to stderr
        handler = logging.StreamHandler(sys.stderr)
        handler.setFormatter(formatter)
        handlers.append(handler)
    elif log_output.startswith("file:"):
        # Log to file
        log_file = log_output[5:]  # Remove 'file:' prefix
        log_path = Path(log_file)

        # Ensure log directory exists
        log_path.parent.mkdir(parents=True, exist_ok=True)

        handler = logging.FileHandler(log_path)
        handler.setFormatter(formatter)
        handlers.append(handler)
    elif log_output == "stdout":
        # Container-friendly: INFO/DEBUG to stdout, WARN/ERROR to stderr
        stdout_handler = logging.StreamHandler(sys.stdout)
        stdout_handler.setFormatter(formatter)
        stdout_handler.setLevel(logging.DEBUG)
        stdout_handler.addFilter(lambda record: record.levelno < logging.WARNING)

        stderr_handler = logging.StreamHandler(sys.stderr)
        stderr_handler.setFormatter(formatter)
        stderr_handler.setLevel(logging.WARNING)

        handlers.extend([stdout_handler, stderr_handler])
    else:
        # Default to stdout for unknown options
        handler = logging.StreamHandler(sys.stdout)
        handler.setFormatter(formatter)
        handlers.append(handler)

    # Configure root logger
    root_logger = logging.getLogger()

    # Clear existing handlers to avoid duplicates
    for handler in root_logger.handlers[:]:
        root_logger.removeHandler(handler)

    # Add new handlers
    for handler in handlers:
        root_logger.addHandler(handler)

    root_logger.setLevel(level)

    # Log startup message
    logger = get_logger(__name__)
    logger.info(f"Logging initialized: level={log_level}, output={log_output}")


def get_logger(name: str) -> logging.Logger:
    """
    Get a logger instance for the specified name.

    Args:
        name: Logger name (typically __name__ from calling module)

    Returns:
        Configured logger instance
    """
    return logging.getLogger(name)


def set_context(**kwargs: Any) -> None:
    """
    Set logging context variables that will be included in all log messages.

    Args:
        **kwargs: Key-value pairs to add to logging context

    Example:
        set_context(operation_id="operation_123", user="admin")
    """
    current_context = _log_context.get({})
    updated_context = {**current_context, **kwargs}
    _log_context.set(updated_context)


def clear_context() -> None:
    """Clear all logging context variables."""
    _log_context.set({})


def update_context(**kwargs: Any) -> None:
    """
    Update existing context with new values.

    Args:
        **kwargs: Key-value pairs to update in logging context
    """
    current_context = _log_context.get({})
    current_context.update(kwargs)
    _log_context.set(current_context)


def get_context() -> Dict[str, Any]:
    """Get current logging context."""
    return _log_context.get({}).copy()


def remove_context_keys(*keys: str) -> None:
    """
    Remove specific keys from logging context.

    Args:
        *keys: Context keys to remove
    """
    current_context = _log_context.get({})
    for key in keys:
        current_context.pop(key, None)
    _log_context.set(current_context)
